In [5]:
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
import pandas as pd
import matplotlib.pyplot as plt

- The average net contribution of a purchase is estimated to be roughly300 NOK. The cost of advertising on the website was 100 NOK for every 1000 ad impressions

- The company also stored additional information about ad exposure (impressions). 

- For the control group, they stored the number of times the user would have been exposed to the ad if they were not in the control group, even though no ad was displayed to them

In [6]:
avg_purchase = 300
ad_cost = 100/1000 # per person

In [7]:
campaign = pd.read_csv('CaseData2025.csv', sep=';')
campaign.head()

test  purchase  impressions
0     0         0            1
1     1         0            1
2     1         0            1
3     1         0            2
4     1         0            6

In [8]:
m = sm.OLS(campaign.purchase, sm.add_constant(campaign.test))
res = m.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               purchase   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     3.401
Date:                Tue, 11 Mar 2025   Prob (F-statistic):             0.0652
Time:                        17:17:01   Log-Likelihood:                 7025.5
No. Observations:               20000   AIC:                        -1.405e+04
Df Residuals:                   19998   BIC:                        -1.403e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0254      0.003      9.360      0.000       0.020       0.031
test           0.0056      0.003      1.844      0.065      -0.000       0.012
==============================================================================
Omnibus:                    20722.183   Durbin-Watson:                   2.003
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           776786.678
Skew:                           5.519   Prob(JB):                         0.00
Kurtosis:                      31.466   Cond. No.                         4.30
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [9]:
m = sm.OLS(campaign.purchase, sm.add_constant(campaign[['test', 'impressions']]))
res = m.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               purchase   R-squared:                       0.031
Model:                            OLS   Adj. R-squared:                  0.031
Method:                 Least Squares   F-statistic:                     323.2
Date:                Tue, 11 Mar 2025   Prob (F-statistic):          7.26e-139
Time:                        17:17:01   Log-Likelihood:                 7341.9
No. Observations:               20000   AIC:                        -1.468e+04
Df Residuals:                   19997   BIC:                        -1.465e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const           0.0134      0.003      4.927      0.000       0.008       0.019
test            0.0047      0.003      1.582      0.114      -0.001       0.011
impressions     0.0045      0.000     25.355      0.000       0.004       0.005
==============================================================================
Omnibus:                    20376.757   Durbin-Watson:                   2.001
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           751879.472
Skew:                           5.368   Prob(JB):                         0.00
Kurtosis:                      31.053   Cond. No.                         23.8
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [10]:
m = smf.ols(formula = "purchase ~ test * impressions", data = campaign)
res = m.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               purchase   R-squared:                       0.033
Model:                            OLS   Adj. R-squared:                  0.033
Method:                 Least Squares   F-statistic:                     230.8
Date:                Tue, 11 Mar 2025   Prob (F-statistic):          3.28e-147
Time:                        17:17:01   Log-Likelihood:                 7364.1
No. Observations:               20000   AIC:                        -1.472e+04
Df Residuals:                   19996   BIC:                        -1.469e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.0207      0.003      7.068      0.000       0.015       0.026
test                -0.0040      0.003     -1.241      0.214      -0.010       0.002
impressions          0.0018      0.000      4.001      0.000       0.001       0.003
test:impressions     0.0033      0.000      6.676      0.000       0.002       0.004
==============================================================================
Omnibus:                    20379.837   Durbin-Watson:                   2.000
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           755396.318
Skew:                           5.367   Prob(JB):                         0.00
Kurtosis:                      31.129   Cond. No.                         34.5
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [11]:
campaign[(campaign.test==1) & (campaign.impressions > 50)].mean()

test            1.000000
purchase        0.459770
impressions    70.804598
dtype: float64

## 1. Is this online advertising campaign effective (i.e., is there an incremental lift in conversions)?

In [12]:
m = smf.ols(formula = "purchase ~ test", data = campaign)
res = m.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               purchase   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     3.401
Date:                Tue, 11 Mar 2025   Prob (F-statistic):             0.0652
Time:                        17:17:01   Log-Likelihood:                 7025.5
No. Observations:               20000   AIC:                        -1.405e+04
Df Residuals:                   19998   BIC:                        -1.403e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0254      0.003      9.360      0.000       0.020       0.031
test           0.0056      0.003      1.844      0.065      -0.000       0.012
==============================================================================
Omnibus:                    20722.183   Durbin-Watson:                   2.003
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           776786.678
Skew:                           5.519   Prob(JB):                         0.00
Kurtosis:                      31.466   Cond. No.                         4.30
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Running a simple OLS were i regress the dummy variable "purchase" on whether the individual were in treatment or control group leads me to believe that the test had an incremental lift in conversions. 

## 2. What is the baseline conversion rate and what is the lift in conversion of the advertising campaign?

The model estimates that 
$$E[Y_i(0)] = 2,54\%, \hspace{5mm} E[Y_i(1)-Y_i(0)] = 0,56\%$$

According to the model the baseline conversion rate is 2.54% and running the ad causes a lift in conversion of 0.56%.

In [16]:
beta = res.params[1]
beta

/var/folders/q8/s4s5342x4pxb1rf55bjmykv40000gn/T/ipykernel_6935/531411608.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  beta = res.params[1]


-0.004040538309619078

## 3. Is the advertising campaign profitable for the company? What would need to change to make the campaign profitable/unprofitable?

In [14]:
print(f'Cost of add per 1000 impressions: {ad_cost*1000}')
print(f'average purchase: {avg_purchase}')
print(f'Lift in conversion: {beta}')
print('------')
print(f'Increase in revenue per 1000 impressions: {round(beta * avg_purchase * 1000,2)}')
print('This is much higher than the cost of ad per 1000 impressions')
print('But one must also account for the fact that one person can see an add multiple times.')

Cost of add per 1000 impressions: 100.0
average purchase: 300
Lift in conversion: 0.0055861056832705565
------
Increase in revenue per 1000 impressions: 1675.83
This is much higher than the cost of ad per 1000 impressions
But one must also account for the fact that one person can see an add multiple times.


## 4. Is there a frequency effect of ad exposure? If yes, how does it affect your conclusions about questions 1-3?

In [15]:
m = smf.ols(formula = "purchase ~ test + test * impressions", data = campaign)
res = m.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               purchase   R-squared:                       0.033
Model:                            OLS   Adj. R-squared:                  0.033
Method:                 Least Squares   F-statistic:                     230.8
Date:                Tue, 11 Mar 2025   Prob (F-statistic):          3.28e-147
Time:                        17:17:01   Log-Likelihood:                 7364.1
No. Observations:               20000   AIC:                        -1.472e+04
Df Residuals:                   19996   BIC:                        -1.469e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.0207      0.003      7.068      0.000       0.015       0.026
test                -0.0040      0.003     -1.241      0.214      -0.010       0.002
impressions          0.0018      0.000      4.001      0.000       0.001       0.003
test:impressions     0.0033      0.000      6.676      0.000       0.002       0.004
==============================================================================
Omnibus:                    20379.837   Durbin-Watson:                   2.000
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           755396.318
Skew:                           5.367   Prob(JB):                         0.00
Kurtosis:                      31.129   Cond. No.                         34.5
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- "test" is the estimate of the treatment effect of individuals that have zero impressions. There is no data in the dataset where <code>impressions = 0</code> and the model cannot be confident that it is non-zero. Obviously the treatment effect of getting zero impressions should be zero. 

- "test:impression" is the interaction term which shows the increase in treatment effect for a unit increase in "impressions", i.e. how much the conversion rate for an individual increases per ad watched. 

This results in changes in the baseline conversion rate, giving a lower estimate than what was previously done. 

## 5. Is there evidence for non-compliance in the data? If so, how would that effect your conclusions?

In [17]:
campaign = pd.read_csv('CaseData2025.csv', sep=';')
campaign.head()

test  purchase  impressions
0     0         0            1
1     1         0            1
2     1         0            1
3     1         0            2
4     1         0            6

In [19]:
campaign.impressions.unique()

array([  1,   2,   6,  35,   3,   7,   5,  22,  72,   4,  23,  15,  17,
        43,  14,  52,   8,  12,  21,  20,  33,  24,  10,  26,   9,  94,
        16,  77,  18,  96,  32,  46,  90,  59,  13,  48,  25,  44,  42,
        74,  11,  19,  69,  78,  63,  49,  28,  37,  27,  30,  62,  51,
        36,  53,  39, 100,  75,  31,  60,  66,  54,  55,  76,  29,  70,
        84,  68,  61,  34,  38,  47,  71,  45,  83,  56,  40,  41,  87,
        86,  99,  50,  89,  88,  57,  67,  98,  93,  58,  73,  92,  91])

## 6. What are important considerations or obstacles the marketing manager should address during future similar experiments? What was good and what could be improved in this experiment?

- Soving an ad offensive and defensive component. By having the treatment group seeing your ad while the control group saw no ad coludes with the internal validity of the experiment. The sole difference between treatment group and control group is not seeing the ad we are testing, but seeing an ad in general. The actions of the control group might have been affected by the fact that they saw no ad, whereas in reality they would likely see another firms ad in place of the one we are testing. This violates SUTVA thorugh the excludability assumption, due to the fact that treatment assignment affects through more than just the treatment variable, which is seeing a specific ad, not a ad in general. 

- As this is an online experiment, it might have issues regarding the interference assumption, due to potential outcomes potentially not beign independent of the treatment status of other individuals. This can be caused by word-of-mouth, control group individualss observing purchases of treatment idividuals, accidentally being exposed to treatment through others devices(or because of deleted cookies).

- It is important not to violate these assumptions as they are key to achieving a unbiased estimate of the ATE.


- There also is ambigious temporal precedence. We know that the cookie is first placed before the first ad. But i do not know from looking at the data when the purchase was made, if purchase variable is 1, whether it was after the first, second, third, etc. Thus the incremental effect of ads might be misleading. 

- Seen as this was an field experiment of sorts, the sample is a good representation of the population and should not be affected by the Hawthorne effect as they are far removed from the experiment. 